feat(#58): add copy to clipboard flag `--copy`

AlejandroSuero opened the pull request:
# Changes made

With this addition the user can choose in interactive mode, or with the flag `--copy`, to copy the created image to their system's clipboard. This was achieved with the [cross-platform clipboard package by golang-design](https://github.com/golang-design/clipboard).

## Tests

- Tested via `make test`:
  - [x] In **MacOS**.
  - [x] In **Windows**.
- [x] Manually tested via `./freeze --config user --lines 23,30 config.go` to get this image:
![image](https://github.com/charmbracelet/freeze/assets/71392160/15a70603-591e-4937-9453-dbd9ffec43f2)

Closes #58.

Moulick commented:
@AlejandroSuero Not working, I cloned your fork, checked out your branch and compiled. `--copy` is not copying to the clipboard. Also `make test` fails. I am on MacOS (14.5 (23F79))

```bash

❯ make test
go test ./...
?   	github.com/charmbracelet/freeze/font	[no test files]
?   	github.com/charmbracelet/freeze/input	[no test files]
?   	github.com/charmbracelet/freeze/svg	[no test files]
?   	github.com/charmbracelet/freeze/test/input	[no test files]
--- FAIL: TestFreezeCopy (0.04s)
    freeze_test.go:81: clipboard is empty
FAIL
FAIL	github.com/charmbracelet/freeze	2.748s
FAIL
make: *** [test] Error 1

```

AlejandroSuero commented:
@Moulick I am using the same version of **MacOS** and it builds correctly with `go build`.

https://github.com/charmbracelet/freeze/assets/71392160/697f3f21-3fd6-481d-9b62-a56359f43c4b

- Result from using `./freeze --config user --lines 23,30 config.go --copy` after being built:
![image](https://github.com/charmbracelet/freeze/assets/71392160/0a291402-2612-4160-8a45-d97c23762073)

As for the test part, I committed a fix, apparently it was reading from the clipboard but not writing it, for example: If I have an image on the clipboard it will pass the test but it will fail otherwise.

- Image from `make test`:
![image](https://github.com/charmbracelet/freeze/assets/71392160/d92183ed-ba8a-4e61-840f-fb9ce61223b3)

Moulick commented:
@AlejandroSuero now the test works, and the image generated by the make test appears in my clipboard, but running `./freeze --lines 23,30 config.go --copy` still writes the image to the disk rather than clipboard. Attaching the compiled binary from my laptop

[freeze.zip](https://github.com/charmbracelet/freeze/files/15481380/freeze.zip)

AlejandroSuero commented:
After `unzip freeze.zip` then `./freeze --execute "exa -l" --copy` I get this (recently powered up MacOS machine with nothing in the clipboard):

![image](https://github.com/charmbracelet/freeze/assets/71392160/d5523170-1ca9-472d-83cc-11097f3cc908)

> But running `./freeze --lines 23,30 config.go --copy` still writes the image to the disk rather than clipboard

With these, are you referring to the fact that it writes the output image as well as copying it to the clipboard?

When running `./freeze --lines 23,30 config.go --copy` with the downloaded [freeze.zip](https://github.com/charmbracelet/freeze/files/15481380/freeze.zip) I get the image `freeze.png` as well as it being copied to the clipboard:

![image](https://github.com/charmbracelet/freeze/assets/71392160/ce6d8338-714c-4c7a-988b-ad080c9c1353)

Moulick commented:
@AlejandroSuero not sure if there is something wrong with my machine but for me `--copy` is writing the `freeze.png` to disk and nothing to clipboard. Maybe we need someone else to try test

AlejandroSuero commented:
https://github.com/charmbracelet/freeze/assets/71392160/e10d653e-c3d0-4c5f-930e-b2a1ef195c39

@Moulick this is how it works for me just powering up the machine and having the clipboard empty.

AlejandroSuero commented on png.go:
```diff
@@ -94,5 +95,12 @@ func resvgConvert(doc *etree.Document, w, h float64, output string) error {
 	if err != nil {
 		return err
 	}
+	if copy {
+		err = clipboard.Init()
+		if err != nil {
+			return err
+		}
+		clipboard.Write(clipboard.FmtImage, png)
+	}
```
I am going to add `clipboard.Read(clipboard.FmtImage)` to see if it works that way for @Moulick.

AlejandroSuero commented:
With the changes in [c57172d](https://github.com/charmbracelet/freeze/pull/97/commits/c57172d67d6c3dba7e1d025b475bbb6794777681), It still passes the tests and works the same way manually for me.

maaslalani commented:
Hey @AlejandroSuero, instead of introducing a new flag I wonder if we should simply allow `--output clipboard` or `--output copy` to copy to the clipboard, what do you think?

meowgorithm commented:
No opinions, but you could also introduce the option to print to stdout so you could `freeze --output stdout | pbcopy`.

maaslalani commented:
> No opinions, but you could also introduce the option to print to stdout so you could `freeze --output stdout | pbcopy`.

In that case I would rather simply detect if stdout is a tty and send to stdout.

```
freeze | pbcopy
```

AlejandroSuero commented:
@maaslalani the problem I contemplated when I thought about it in the first place was that if for example you want to write an output name like `artichoke_example.png` and still you want to copy the image to the clipboard it won't be possible.

But I would like to hear your thoughts about that scenario.

AlejandroSuero commented:
@meowgorithm as for the `stdout` @Moulick mentioned this in the issue linked to this PR:

- https://github.com/charmbracelet/freeze/issues/58#issuecomment-2120841624

Moulick commented:
1. It looks like something is broken on my system, as I ran a new MacOS VM and it works fine there. I have no clue what's broken on my system but probably a myriad of third-party applications I am running that might be interfering with the clipboard. ¯\\\_(ツ)_/¯ 
2. `--output clipboard` does sound like a better idea. I don't think anyone want to copy to clipboard _and_ write to disk at the same time. Even on Android/iOS, the option when taking a screenshot is copy to clipboard and delete.
3. `pbcopy` does not handle putting images into the clipboard correctly. So `freeze | pbcopy` cannot work.
    1. `freeze --output stdout` or detecting stdout tty should also be done, so we can use other programs (like image compression) in the workflow but probably in a separate PR.

AlejandroSuero commented:
@maaslalani @Moulick lets see if you can help me figure out why `freeze --output clipboard isn't working.

I made the following changes:

- On `png.go`, refactored `if copyClipboard` for:

```go
  switch output {
  case "clipboard":
    err = clipboard.Init()
    if err != nil {
      return err
    }
    clipboard.Write(clipboard.FmtImage, png)
    clipboard.Read(clipboard.FmtImage)
  break
  default:
    err = os.WriteFile(output, png, 0644)
    if err != nil {
      return err
    }
  }
```

But it does not copy to clipboard it just writes file `clipboard`.

<img width="661" alt="Screenshot 2024-06-02 at 16 34 56" src="https://github.com/charmbracelet/freeze/assets/71392160/47beb1db-6a8e-4175-a53b-31051a634799">

> [!NOTE]
> I removed all instances of `config.Copy`

AlejandroSuero commented:
@Moulick I made some changes to work with `librsvg` and for non-png formats.

I added some comments on how [gclip](https://github.com/golang-design/clipboard) does not seem to work as an image when copying non-png formats. I wrote an issue about it in their github to figure it out.

- https://github.com/golang-design/clipboard/issues/58

AlejandroSuero commented:
As @maaslalani suggested, I made the changes so the user can `freeze cut.go --output clipboard`.

With these changes (9a169d5, 01d684a), the comment https://github.com/charmbracelet/freeze/pull/97#issuecomment-2162657455 is worked around so it will always use `.png` format to copy the image.

<details>
<summary>Test Results</summary>

```bash
> go test -v ./...
?       github.com/charmbracelet/freeze/font    [no test files]
?       github.com/charmbracelet/freeze/input   [no test files]
?       github.com/charmbracelet/freeze/test/input      [no test files]
?       github.com/charmbracelet/freeze/svg     [no test files]
=== RUN   TestConfig
--- PASS: TestConfig (0.00s)
=== RUN   TestCut
--- PASS: TestCut (0.00s)
=== RUN   TestFreeze
--- PASS: TestFreeze (0.27s)
=== RUN   TestFreezeOutput
--- PASS: TestFreezeOutput (0.04s)
=== RUN   TestFreezeCopy
--- PASS: TestFreezeCopy (0.50s)
=== RUN   TestFreezeHelp
--- PASS: TestFreezeHelp (0.01s)
=== RUN   TestFreezeErrorFileMissing
--- PASS: TestFreezeErrorFileMissing (0.01s)
=== RUN   TestFreezeConfigurations
=== RUN   TestFreezeConfigurations/artichoke-base
=== RUN   TestFreezeConfigurations/artichoke-full
=== RUN   TestFreezeConfigurations/eza
=== RUN   TestFreezeConfigurations/execute
=== RUN   TestFreezeConfigurations/bubbletea
=== RUN   TestFreezeConfigurations/bubbletea-copy
=== RUN   TestFreezeConfigurations/layout
=== RUN   TestFreezeConfigurations/haskell
=== RUN   TestFreezeConfigurations/dracula
=== RUN   TestFreezeConfigurations/border-radius
=== RUN   TestFreezeConfigurations/window
=== RUN   TestFreezeConfigurations/border-width
=== RUN   TestFreezeConfigurations/padding
=== RUN   TestFreezeConfigurations/margin
=== RUN   TestFreezeConfigurations/shadow
=== RUN   TestFreezeConfigurations/dimensions
=== RUN   TestFreezeConfigurations/dimensions-margin
=== RUN   TestFreezeConfigurations/dimensions-margin-line-numbers
=== RUN   TestFreezeConfigurations/dimensions-padding
=== RUN   TestFreezeConfigurations/dimensions-config
=== RUN   TestFreezeConfigurations/overflow
=== RUN   TestFreezeConfigurations/lines
=== RUN   TestFreezeConfigurations/font-size-28
=== RUN   TestFreezeConfigurations/font-size-14
=== RUN   TestFreezeConfigurations/line-height-2
=== RUN   TestFreezeConfigurations/overflow-line-numbers
=== RUN   TestFreezeConfigurations/helix
=== RUN   TestFreezeConfigurations/glow
=== RUN   TestFreezeConfigurations/tab
--- PASS: TestFreezeConfigurations (0.77s)
    --- PASS: TestFreezeConfigurations/artichoke-base (0.02s)
    --- PASS: TestFreezeConfigurations/artichoke-full (0.02s)
    --- PASS: TestFreezeConfigurations/eza (0.02s)
    --- PASS: TestFreezeConfigurations/execute (0.03s)
    --- PASS: TestFreezeConfigurations/bubbletea (0.03s)
    --- PASS: TestFreezeConfigurations/bubbletea-copy (0.03s)
    --- PASS: TestFreezeConfigurations/layout (0.03s)
    --- PASS: TestFreezeConfigurations/haskell (0.03s)
    --- PASS: TestFreezeConfigurations/dracula (0.03s)
    --- PASS: TestFreezeConfigurations/border-radius (0.03s)
    --- PASS: TestFreezeConfigurations/window (0.03s)
    --- PASS: TestFreezeConfigurations/border-width (0.03s)
    --- PASS: TestFreezeConfigurations/padding (0.03s)
    --- PASS: TestFreezeConfigurations/margin (0.03s)
    --- PASS: TestFreezeConfigurations/shadow (0.03s)
    --- PASS: TestFreezeConfigurations/dimensions (0.03s)
    --- PASS: TestFreezeConfigurations/dimensions-margin (0.03s)
    --- PASS: TestFreezeConfigurations/dimensions-margin-line-numbers (0.03s)
    --- PASS: TestFreezeConfigurations/dimensions-padding (0.03s)
    --- PASS: TestFreezeConfigurations/dimensions-config (0.03s)
    --- PASS: TestFreezeConfigurations/overflow (0.03s)
    --- PASS: TestFreezeConfigurations/lines (0.03s)
    --- PASS: TestFreezeConfigurations/font-size-28 (0.03s)
    --- PASS: TestFreezeConfigurations/font-size-14 (0.03s)
    --- PASS: TestFreezeConfigurations/line-height-2 (0.03s)
    --- PASS: TestFreezeConfigurations/overflow-line-numbers (0.03s)
    --- PASS: TestFreezeConfigurations/helix (0.03s)
    --- PASS: TestFreezeConfigurations/glow (0.02s)
    --- PASS: TestFreezeConfigurations/tab (0.02s)
PASS
ok      github.com/charmbracelet/freeze 3.405s
```

</details>

bashbunni commented:
Hey just an update on this - I want to see how we can improve where this fits in the codebase. The feature works great, I just think we can probably refactor a bit to leave the codebase cleaner than we found it. 

I also think `--output clipboard` is better nomenclature than `--output copy` so I removed that option in my update here. Sorry, this PR will sit a bit longer to figure out the cleanest solution 🙏 Thanks for your patience

AlejandroSuero commented:
@bashbunni I will try and look into it this weekend and see how it turns out, thanks for the feedback.

winterjung commented:
Is there any update on this PR? Currently, I'm running `freeze ...`, then `open freeze.png` in terminal, and manually copying it to the clipboard using <kbd>command</kbd> + <kbd>c</kbd>.

Really looking forward to seeing this PR merged soon — it would make the workflow so much smoother!

caarlos0 commented on png.go:
```diff
@@ -5,12 +5,24 @@ import (
+		return err
+	}
+	clipboard.Write(clipboard.FmtImage, img)
+	clipboard.Read(clipboard.FmtImage)
```
i don't think you need this read here

AlejandroSuero replied:
From when I was testing it, it seemed that if the `.Read` is removed it will error because it thinks it's empty. But I will try and test it this weekend again.

caarlos0 commented on main.go:
```diff
@@ -394,6 +394,9 @@ func main() {
 
 	istty := isatty.IsTerminal(os.Stdout.Fd())
 
+	if config.Output == "clipboard" { // convert to png because we can't copy svg to clipboard
+		config.Output = "clipboard.png"
+	}
```
shouldn't we then copy the svg as a text file instead?

AlejandroSuero replied:
I can try do it that way, but if you want to copy the image to share when pasting, for example in a GitHub comment or in another app. If it is a text file, it won't recognise it as an image. Maybe if pasting it in a web browser to open the image there it will work.

It can be checked if the desired output is `.svg` that it copies it as text. But in the case that you want to share the image, it will have to be done as mentioned above.

caarlos0 replied:
I mean, you could still paste the SVG into a html file for example... not sure about other places

AlejandroSuero replied:
Yeah, I get your point, it's useful for that case. I'll look into adding the copy as text when the desired output is `.svg`.

bashbunni replied:
No I think it should be png. Save to file for svg since that's most likely to be used with another program or in cases where we need a file to exist for it (e.g. HTML, or use in Figma, Illustrator, etc). Png is a more convenient format that lends itself well to being pasted _anywhere_.
